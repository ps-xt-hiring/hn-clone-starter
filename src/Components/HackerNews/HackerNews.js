/* eslint-disable */
import React, { useEffect, Fragment } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { getHitsData, upVote, hideFeed } from '../../store/Actions/actions';
import * as S from './HackerNews.style';

import Feed from '../Feed';

const HackerNews = () => {
  const feedData = useSelector(state => state.reducer);
  const dispatch = useDispatch();
  const {
    data: hits,
    totalPages,
    apiRequestType,
    loading,
  } = feedData;
  let {
    page: numOfPage,
  } = feedData;
  let showMore = false;

  const hiddenFeeds = localStorage.getItem('hiddenFeeds')
    ? JSON.parse(localStorage.getItem('hiddenFeeds')) : [];

  const getInitialData = () => (
    dispatch(getHitsData('top', numOfPage))
  );

  useEffect(() => {
    getInitialData();
  }, []);

  const upVoteFeed = (feedId) => {
    dispatch(upVote(feedId));
  };

  const hideFeedFromView = (feedId) => {
    dispatch(hideFeed(feedId));
  };

  const showMoreFeed = () => {
    numOfPage += 1;
    dispatch(getHitsData(apiRequestType, numOfPage));
  };

  if (hits && hits.length) {
    hits.map((feed) => {
      feed.isHidden = hiddenFeeds.length && hiddenFeeds.indexOf(feed.objectID) > -1;
    });
  }

  showMore = numOfPage < totalPages;

  return (
    <Fragment>
      {!loading ? (
        <div>
          <S.StyledUl>
            {hits && hits.length && (
              hits.map((feed) => {
                return (
                  !feed.isHidden && (
                    <Feed
                      feed={feed}
                      key={feed.objectID}
                      upVoteFeed={upVoteFeed}
                      hideFeed={hideFeedFromView}
                    />
                  )
                )
              })
            )}
          </S.StyledUl>
          {showMore && (
            <S.ShowMore onClick={() => showMoreFeed()}>More</S.ShowMore>
          )}
        </div>
      ) : (
        <S.Loading>Loading...</S.Loading>
      )}
    </Fragment>
  );
};

HackerNews.propTypes = {
  feedData: PropTypes.shape({
    data: PropTypes.array,
    totalPages: PropTypes.number,
    apiRequestType: PropTypes.string,
    loading: PropTypes.bool,
  }),
};

HackerNews.defaultProps = {
  feedData: {},
};

export default HackerNews;
