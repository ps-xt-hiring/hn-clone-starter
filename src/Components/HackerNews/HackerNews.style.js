/* eslint-disable */
import styled from 'styled-components';
import Colors from '../../constants/theme';

export const Loading = styled.div`
  font-size: 18px;
  text-align: center;
  padding: 10px;
`;

export const StyledUl = styled.ul`
  background-color: ${Colors.list};
  padding: 0;
`;

export const ShowMore = styled.div`
  color: ${Colors.showMore};
  text-align: center;
  padding-bottom: 15px;
  cursor: pointer;

  &:hover {
    text-decoration: underline;
  }
`;
