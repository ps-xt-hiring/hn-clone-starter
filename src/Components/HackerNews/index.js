/* eslint-disable */
export { default } from './HackerNews';
