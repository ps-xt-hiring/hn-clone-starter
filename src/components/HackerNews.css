
.HackerNewsWrapper {
  background: #f6f6ef;
  width: 85%;
  margin: 0 auto;
}

@media only screen and  (max-width: 768px) {
  .HackerNewsWrapper {
    width: 96%;
  }
}

.HackerNewsWrapper h1 {
  font-size: 20px;
  font-weight: 300;
  color: #222;
  background: #f60;
}

