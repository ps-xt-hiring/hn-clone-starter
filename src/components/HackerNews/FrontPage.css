ul {
  font-size: small;
  color: black;
}

li {
  list-style: none;
  padding: 8px 0;
  display: flex;
  align-items: center;
}

li:nth-child(odd) {
  background-color: whitesmoke;
}

li:nth-child(even) {
  background-color: rgba(90, 86, 86, 0.2);
}

.comments {
  font-size: 10pt;
  flex: 1 1;
  text-align: end;
}

.details {
  flex: 12 1;
  margin-left: 16px;
}
