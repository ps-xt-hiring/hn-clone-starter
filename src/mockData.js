export const HackerNewsData = {
  isFetching: false,
  error: null,
  pageNo: 1,
  news:[
    {
  created_at: "2016-02-17T08:38:37.000Z",
  title: "A Message to Our Customers",
  url: "http://www.apple.com/customer-letter/",
  author: "epaga",
  points: 5771,
  story_text: null,
  comment_text: null,
  num_comments: 967,
  story_id: null,
  story_title: null,
  story_url: null,
  parent_id: null,
  created_at_i: 1455698317,
  relevancy_score: 6561,
  _tags:  ["story", "author_epaga", "story_11116274"],
  objectID: "11116274"
  }],
  updateVotesAction: () => {},
}

export const listWrapperData = {
  isFetching: false,
  error: null,
  pageNo: 1,
  news:[
    {
  created_at: "2016-02-17T08:38:37.000Z",
  title: "A Message to Our Customers",
  url: "http://www.apple.com/customer-letter/",
  author: "epaga",
  points: 5771,
  story_text: null,
  comment_text: null,
  num_comments: 967,
  story_id: null,
  story_title: null,
  story_url: null,
  parent_id: null,
  created_at_i: 1455698317,
  relevancy_score: 6561,
  _tags:  ["story", "author_epaga", "story_11116274"],
  objectID: "11116274"
  },
  {
    created_at: "2016-02-17T08:38:37.000Z",
    title: "A Message to Our Customers",
    url: "http://www.apple.com/customer-letter/",
    author: "epaga",
    points: 5771,
    story_text: null,
    comment_text: null,
    num_comments: 967,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1455698317,
    relevancy_score: 6561,
    _tags:  ["story", "author_epaga", "story_11116274"],
    objectID: "11116274"
    }
],
  updateVotesAction: () => {},
  fetchNewsForFirstPage: () =>{}
};

export const listItemData = {
  index: 1,
  item:{
  created_at: "2016-02-17T08:38:37.000Z",
  title: "A Message to Our Customers",
  url: "http://www.apple.com/customer-letter/",
  author: "epaga",
  points: 5771,
  story_text: null,
  comment_text: null,
  num_comments: 967,
  story_id: null,
  story_title: null,
  story_url: null,
  parent_id: null,
  created_at_i: 1455698317,
  relevancy_score: 6561,
  _tags:  ["story", "author_epaga", "story_11116274"],
  objectID: "11116274"
  },
  updateVotesAction: () => {}
};
