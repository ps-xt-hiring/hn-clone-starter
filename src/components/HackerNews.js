import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ListWrapper from './ListWrapper/ListWrapper';
import './HackerNews.css';

class HackerNews extends Component {
  static defaultProps = {
    pageNo: 0,
  };

  static propTypes = {
    pageNo: PropTypes.number.isRequired,
    isFetching: PropTypes.bool.isRequired,
    fetchNewsForFirstPage: PropTypes.func.isRequired,
    error: PropTypes.object,
  };

  componentDidMount() {
    const { pageNo, fetchNewsForFirstPage } = this.props;
    fetchNewsForFirstPage(pageNo);
  }

  render() {
    const { error, isFetching } = this.props;
    if (error) {
      return <div>Error: {error.message}</div>;
    } else if (isFetching) {
      return <div>Loading...</div>;
    }
    return (
      <div className="HackerNewsWrapper">
        <h1>Hacker News</h1>
        <ListWrapper {...this.props} />
      </div>
    );
  }
}

export default HackerNews;
