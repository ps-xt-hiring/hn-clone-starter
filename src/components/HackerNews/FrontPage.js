import React, { Component } from 'react';
import './FrontPage.css';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as hackerNewsActions from '../../api/services';
import HNHeader from './HN_Header/HNHeader';
import HNFooter from './HN_Footer/HNFooter';
import HNRow from './HN_Row/HNRow';

class FrontPage extends Component {
  state = {
    hnlist: [],
  };

  componentDidMount() {
    const { hackerNewsDispatchActions } = this.props;
    hackerNewsDispatchActions.getHackerNewsData();
  }

  componentWillReceiveProps(nextProps) {
    const { hnList } = nextProps.hackerNewsObject;
    this.setState({ hnlist: hnList });
  }

  hideHNRowHandler = (objectID) => {
    const { hnlist } = this.state;
    const { hackerNewsDispatchActions } = this.props;
    hackerNewsDispatchActions.removeObjFromHNData(hnlist, objectID);
  };

  upVoteHNewsHandler = (objectID) => {
    const { hnlist } = this.state;
    const { hackerNewsDispatchActions } = this.props;
    hackerNewsDispatchActions.upVoteHNews(hnlist, objectID);
  };

  render() {
    const { hackerNewsDispatchActions, hackerNewsObject } = this.props;
    const { getHackerNewsData } = hackerNewsDispatchActions;
    const { loading, currentPage, totalPage } = hackerNewsObject;
    const { hnlist } = this.state;
    return (
      <div>
        <HNHeader getHackerNewsData={getHackerNewsData} />

        <div className="hn-contents">
          <ul>
            {!loading
              && hnlist
              && hnlist.map(item => (
                <HNRow
                  key={JSON.stringify(item)}
                  {...item}
                  hideHNRowHandler={this.hideHNRowHandler}
                  upVoteHNewsHandler={this.upVoteHNewsHandler}
                />
              ))}
            {loading && <span>Loading...</span>}
            <li>
              <HNFooter
                actiondata={{ getHackerNewsData, currentPage, totalPage }}
              />
            </li>
          </ul>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  hackerNewsObject: state.hackerNews,
});
const mapDispatchToProps = dispatch => ({
  hackerNewsDispatchActions: bindActionCreators(hackerNewsActions, dispatch),
});
export default connect(mapStateToProps, mapDispatchToProps)(FrontPage);
