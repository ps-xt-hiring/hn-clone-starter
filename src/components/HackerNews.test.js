import React from 'react';
import { shallow } from 'enzyme';
import HackerNews from './HackerNews';
import { HackerNewsData } from '../mockData';
const { pageNo, isFetching, error, news, updateVotesAction } = HackerNewsData;

describe('HackerNews Component', () => {
  let wrapper;
  const mockfetchNewsFirstPage = jest.fn();
  beforeEach(() => {
    mockfetchNewsFirstPage.mockClear();
    wrapper = shallow(
      <HackerNews
        fetchNewsForFirstPage={mockfetchNewsFirstPage}
        isFetching={isFetching}
        pageNo={pageNo}
        error={error}
        news={news}
        updateVotesAction={updateVotesAction}
      />
    );
  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the mock fetchNewsForFirstPage function on Mount', () => {
    expect(mockfetchNewsFirstPage.mock.calls.length).toBe(1);
  });
});
