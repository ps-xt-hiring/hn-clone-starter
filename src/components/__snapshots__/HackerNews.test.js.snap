// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackerNews Component renders without crashing 1`] = `ShallowWrapper {}`;
